docs/extensions/storage-transformers/sharding/v1.0.rst

```diff
@@ -0,0 +1,293 @@
+``chunk separator``: For all entries that are part of the same shard the key is
+changed to the shard-key and the values are combined in the
+`Binary shard format`_ as described below. The new shard-key is the chunk key
+divided by ``chunks_per_shard`` and floored per dimension. For example for
+``chunks_per_shard=[32, 2]``, the chunk grid position ``[96, 18]`` (e.g. key
+"data/root/foo/baz/c96/18") is transformed to the shard grid position
+``[3, 9]`` and reassigned to the respective new key, honoring the original chunk
+separator (e.g. "data/root/foo/baz/c3/9"). Chunk grid positions ``[96, 19]``,
+``[97, 18]``, …, up to ``[127, 19]`` will also have the same shard grid position
+``[3, 9]``.
+
+
+Binary shard format
+-------------------
+
+The only binary format is the ``indexed`` format, as specified by the ``type``
+configuration key. Other binary formats might be added in future versions.
+
+In the indexed binary format, chunks are written successively in a shard, where
+unused space between them is allowed, followed by an index referencing them. The
+index is placed at the end of the file and has a size of 16 bytes multiplied by
+the number of chunks in a shard, for example ``16 bytes * 64 = 1014 bytes`` for
+``chunks_per_shard=[32, 2]``. The index holds an `offset, nbytes` pair of
+little-endian uint64 per chunk, the chunks-order in the index is row-major (C)
+order, for example for ``chunks_per_shard=[2, 2]`` an index would look like:
+
+.. code-block::
+
+    | chunk (0, 0)    | chunk (0, 1)    | chunk (1, 0)    | chunk (1, 1)    |
+    | offset | nbytes | offset | nbytes | offset | nbytes | offset | nbytes |
+    | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 | uint64 |
+
+
+Empty chunks are denoted by setting both offset and nbytes to ``2^64 - 1``. The
+index always has the full shape of all possible chunks per shard, even if they
+are outside of the array size.
+
+The actual order of the chunk content is not fixed and may be chosen by the
+implementation as all possible write orders are valid according to this
+specification and therefore can be read by any other implementation. When
+writing partial chunks into an existing shard no specific order of the existing
+chunks may be expected. Some writing strategies might be
+
+* **Fixed order**: Specify a fixed order (e.g. row-, column-major, or Morton
+  order). When replacing existing chunks larger or equal-sized chunks may be
+  replaced in-place, leaving unused space up to an upper limit that might
+  possibly be specified. Please note that for regular-sized uncompressed data
+  all chunks have the same size and can therefore be replaced in-place. > *
+* **Append-only**: Any chunk to write is appended to the existing shard,
+  followed by an updated index. If previous chunks are updated, their storage
+  space becomes unused, as well as the previous index. This might be useful for
+  storage that only allows append-only updates.
```

Comment:
This is starting to sound a bit similar to how TileDB works. 

Would an implementation want to support compaction in this case?

Reply:
I guess that's up to the implementation(s). For the moment I think it's not necessary to specify this in the spec. Also those strategies are just examples to get an idea what *could* be implemented. In my opinion, it's most important that all clients can read the sharded arrays, no matter what strategy was used for writing them.